frame/session/src/lib.rs

```diff
@@ -498,36 +499,23 @@ decl_module! {
 				Self::rotate_session();
 			}
 		}
-
-		/// Called when the runtime is upgraded.
-		fn on_runtime_upgrade() {
-			Self::migrate();
-		}
 	}
 }
 
-impl<T: Trait> Module<T> {
-	/// Move keys from NextKeys and KeyOwner, if any exist.
-	fn migrate() {
-		use frame_support::storage::migration::{put_storage_value, StorageIterator};
-		sp_runtime::print("Migrating session's double-maps...");
-
-		let prefix = {
-			const DEDUP_KEY_PREFIX: &[u8] = b":session:keys";
-			let encoded_prefix_key_hash = codec::Encode::encode(&DEDUP_KEY_PREFIX);
-			let mut h = sp_io::hashing::twox_64(&encoded_prefix_key_hash[..]).to_vec();
-			h.extend(&encoded_prefix_key_hash[..]);
-			h
-		};
-
-		for (hash, value) in StorageIterator::<T::Keys>::with_suffix(b"Session", b"NextKeys", &prefix[..]).drain() {
-			put_storage_value(b"Session", b"NextKeys", &hash, value);
-		}
-		for (hash, value) in StorageIterator::<T::ValidatorId>::with_suffix(b"Session", b"KeyOwner", &prefix[..]).drain() {
-			put_storage_value(b"Session", b"KeyOwner", &hash, value);
+impl<T: Trait> MigrateAccount<T::AccountId> for Module<T> {
+	fn migrate_account(a: &T::AccountId) {
+		use codec::Encode;
+		if let Ok(v) = a.using_encoded(|mut d| T::ValidatorId::decode(&mut d)) {
```

Comment:
This should use `T::ValidatorIdOf::convert(a.clone())`. The code as-is attempts to interpret an account key as a `ValidatorId` (in practice - stash key) where what we really want to do is look up the stash key by account ID in the staking module. This is what `ValidatorIdOf` does in practice.